File: Structures/RedBlackTree/rbTree.go
package main

////////////////////////////////
//    Author : Armin Goodarzi //
//    GitHub :                //
//      github.com/Armingodiz //
//                            //
////////////////////////////////

import (
	"fmt"
	"github.com/fatih/color"
	"strconv"
)

///////////////////////////////////////// types :

type ColorNode int

const (
	Black ColorNode = iota
	Red
)

type Node struct {
	Parent *Node
	Left   *Node
	Right  *Node
	Color  ColorNode
	key    int
}

type Tree struct {
	Root  *Node
	dummy *Node
}

type TreePicture struct {
	pic string
}

//////////////////////////////////////////////

/////////////////////////////////////// create funcs :

func (tree *Tree) CreateNode(key int) *Node {
	return &Node{
		Parent: nil,
		Left:   tree.dummy,
		Right:  tree.dummy,
		Color:  Red,
		key:    key,
	}
}

func CreateTree() *Tree {
	tree := &Tree{
		Root:  nil,
		dummy: nil,
	}
	tree.dummy = tree.CreateNode(-1234)
	tree.dummy.Color = Black
	return tree
}

/////////////////////////////////////////////

////////////////////////////////////////////////////////// getter funcs :

func (node *Node) getUncle() *Node {
	if node.Parent.IsRightChild() {
		return node.Parent.Parent.Left
	}
	return node.Parent.Parent.Right
}

func (node *Node) getGrandParent() *Node {
	return node.Parent.Parent
}

/////////////////////////////////////////////////////////////////////////

///////////////////////////////////////////////////////////// other funcs :

func (node *Node) Search(tree *Tree, key int) *Node {
	if node == tree.dummy {
		fmt.Println("KEY DOES NOT EXIST !")
		return nil
	}
	if node.key == key {
		return node
	}
	if key > node.key {
		return node.Right.Search(tree, key)
	} else {
		return node.Left.Search(tree, key)
	}
}

func (node *Node) IsRightChild() bool {
	if node.key > node.Parent.key {
		return true
	}
	return false
}

/////////////////////////////////////////////////////////////////////////

//////////////////////////////////////////////////// Rotations funcs :

func (tree *Tree) Rotate_right(node *Node) {
	// setting targetNode.right as node.left
	// node as and targetNode.right
	targetNode := node.Left
	node.Left = targetNode.Right
	node.Left.Parent = node
	targetNode.Right = node
	setParrentsAfterRotation(tree, node, targetNode)
}

func (tree *Tree) Rotate_left(node *Node) {
	// setting targetNode.Left as node.Right
	// and setting node as targetNode.Left
	targetNode := node.Right
	node.Right = targetNode.Left
	node.Right.Parent = targetNode
	targetNode.Left = node
	setParrentsAfterRotation(tree, node, targetNode)
}

func setParrentsAfterRotation(tree *Tree, node, targetNode *Node) {
	// setting targetNode.parent
	if node == tree.Root { //node was the root of the tree
		tree.Root = targetNode
		targetNode.Parent = nil
	} else {
		targetNode.Parent = node.Parent
		if node.IsRightChild() {
			node.Parent.Right = targetNode
		} else {
			node.Parent.Left = targetNode
		}
	}
	// setting targetNode as node's parent
	node.Parent = targetNode
}

////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////// insertion :

func (tree *Tree) Insert(data int) {
	newNode := tree.CreateNode(data)
	if tree.Root == nil {
		tree.Root = newNode
		tree.Root.Color = Black
	} else {
		tmp := tree.Root
		for {
			holder := tmp
			if data > tmp.key {
				tmp = tmp.Right
				if tmp == tree.dummy {
					holder.Right = newNode
					newNode.Parent = holder
					break
				}
			} else {
				tmp = tmp.Left
				if tmp == tree.dummy {
					holder.Left = newNode
					newNode.Parent = holder
					break
				}
			}
		}
		tree.fixRbViolations(newNode)
	}
}

func (tree *Tree) fixRbViolations(node *Node) {
	for node != tree.Root && node.Color == Red && node.Parent.Color == Red {
		if !node.Parent.IsRightChild() { // for LL and LR cases :
			if node.getUncle() != nil && node.getUncle().Color == Red { // case L.1 :
				node.Parent.Color = Black
				node.getUncle().Color = Black
				node.getGrandParent().Color = Red
				node = node.getGrandParent()
			} else { // case L.2 & L.3
				if node.IsRightChild() { // case L.2 = LR
					node = node.Parent
					tree.Rotate_left(node)
				} //case L.3 = LL :
				node.Parent.Color = Black
				node.getGrandParent().Color = Red
				tree.Rotate_right(node.getGrandParent())
			}
		} else { // for RR and RL cases :
			if node.getUncle() != nil && node.getUncle().Color == Red { // case R.1 :
				node.Parent.Color = Black
				node.getUncle().Color = Black
				node.getGrandParent().Color = Red
				node = node.getGrandParent()
			} else { // case R.2 & R.3
				if !node.IsRightChild() { // case R.2 = RL
					node = node.Parent
					tree.Rotate_right(node)
				} //case R.3 = RR :
				node.Parent.Color = Black
				node.getGrandParent().Color = Red
				tree.Rotate_left(node.getGrandParent())
			}
		}
	}
	tree.Root.Color = Black
}

/////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////// TODOs :
//TODO: implement deletion

/////////////////////////////////////////////////////////////////// displaying tree :

func GetTreePic(tree *Tree, res *TreePicture, padding string, pointer string, node *Node) {
	if node != tree.dummy {
		res.pic = res.pic + padding
		res.pic = res.pic + pointer
		key := strconv.Itoa(node.key)
		if node.Color == Red {
			red := color.New(color.FgRed).SprintFunc()
			key = red(key)
		}
		res.pic = res.pic + key
		res.pic = res.pic + "\n"

		paddingBuilder := padding + "│  "
		paddingForBoth := paddingBuilder
		pointerForRight := "└──"
		pointerForLeft := ""
		if node.Right != nil {
			pointerForLeft = "├──"
		} else {
			pointerForLeft = "└──"
		}

		GetTreePic(tree, res, paddingForBoth, pointerForLeft, node.Left)
		GetTreePic(tree, res, paddingForBoth, pointerForRight, node.Right)
	}
}

func (tree *Tree) DisplayTree() {
	pic := &TreePicture{pic: ""}
	GetTreePic(tree, pic, "", "", tree.Root)
	fmt.Println(pic.pic)
}

/////////////////////////////////////////////////////////////////////////////////

///////////////////////////////////////////////////////// testing implemention :
func main() {
	tree := CreateTree()
	tree.Insert(10)
	tree.DisplayTree()
	fmt.Println("######################################")
	tree.Insert(18)
	tree.DisplayTree()
	fmt.Println("######################################")
	tree.Insert(7)
	tree.DisplayTree()
	fmt.Println("######################################")
	tree.Insert(15)
	tree.DisplayTree()
	fmt.Println("######################################")
	tree.Insert(16)
	tree.DisplayTree()
	fmt.Println("######################################")
	tree.Insert(30)
	tree.DisplayTree()
	fmt.Println("######################################")
	tree.Insert(25)
	tree.DisplayTree()
	fmt.Println("######################################")
	tree.Insert(40)
	tree.DisplayTree()
	fmt.Println("######################################")
	tree.Insert(60)
	tree.DisplayTree()
	fmt.Println("######################################")
	tree.Insert(2)
	tree.DisplayTree()
	fmt.Println("######################################")
	tree.Insert(1)
	tree.DisplayTree()
	fmt.Println("######################################")
	tree.Insert(70)
	tree.DisplayTree()
	fmt.Println("######################################")
}

File: Structures/RedBlackTree/CPP/RBTree.cpp
#include <iostream>
using namespace std;
enum Color {red,black};
class Node{
    public :
    Node* parent;
    Node* left;
    Node* right;
    Color color;
    int key;
    Node(int key,Color color){
        this->key=key;
        parent=left=right=NULL;
        this->color=color;
    }
};
class RBTree{
    private:
        Node* root;
        Node* nullNode;
        Node* flipColors(Node* parentOfParent){
            parentOfParent->color=red;
            parentOfParent->right->color=black;
            parentOfParent->left->color=black;
            return parentOfParent;
        }
        Node* rotateRight(Node* node){     
            Node* leftNode=node->left;
            node->left=leftNode->right;
            if(leftNode->right!=nullNode)leftNode->right->parent=node;
            leftNode->parent=node->parent;
            if(node->parent==nullNode)
                root=leftNode;
            else{
                if(node==node->parent->right)node->parent->right=leftNode;
                else node->parent->left=leftNode;
            }
            leftNode->right=node;
            node->parent=leftNode;
            return leftNode;
        }
        Node* rotateLeft(Node* node){
            Node* rightNode=node->right;
            node->right=rightNode->left;
            if(rightNode->left!=nullNode)rightNode->left->parent=node;
            rightNode->parent=node->parent;
            if(node->parent==nullNode)
                root=rightNode;
            else{
                if(node==node->parent->right)node->parent->right=rightNode;
                else node->parent->left=rightNode;
            }
            rightNode->left=node;
            node->parent=rightNode;
            return rightNode;
        }
        Node* minNode(Node* currentNode){
            if (currentNode==nullNode)return nullNode;
            while(currentNode->left!=nullNode){
                currentNode=currentNode->left;
            }
            return currentNode;
        }
        Node* inorder_successor(Node* node){
            if (node==nullNode)return nullNode;
            if(node->right!=nullNode)return minNode(node->right);
            Node* parent=node->parent;
            while(parent!=nullNode && node==parent->right){
                node=parent;
                parent=parent->parent;
            }
            return parent;
        }
    public:
        RBTree(){
            nullNode=new Node(-1,black);
            root=nullNode;

        }
        void RB_insert(int key){
        Node* newNode=create_node(key);
        Node* parent=nullNode;
        Node* currentNode=root;
        while (currentNode!=nullNode){
            parent=currentNode;
            currentNode= newNode->key>currentNode->key?
                currentNode->right:currentNode->left;
        }
        newNode->parent=parent;
        if(parent==nullNode)root=newNode;
        else{
            if(newNode->key<parent->key)parent->left=newNode;
            else parent->right=newNode;
        }
            RB_insert_fixup(newNode);
        }
        void RB_insert_fixup(Node* node){
            while(node->parent->color==red){
                Node* parentNode=node->parent;
                if(parentNode==parentNode->parent->left){
                    Node* uncle=parentNode->parent->right;
                    if(uncle->color=red){
                        node=flipColors(uncle->parent);//case 1
                    }else{
                        if(node==parentNode->right){//case 2
                            node=parentNode;
                            rotateLeft(node);
                        }
                        //case 3
                        node->parent->color=black;
                        node->parent->parent->color=red;
                        rotateRight(node->parent->parent);
                    }
                }
                else{
                    Node* uncle=parentNode->parent->left;
                    if(uncle->color=red){
                        node=flipColors(uncle->parent);//case 1
                    }else{
                        if(node==parentNode->left){//case 2
                            node=parentNode;
                            rotateRight(node);
                        }
                        //case 3
                        node->parent->color=black;
                        node->parent->parent->color=red;
                        rotateLeft(node->parent->parent);
                    }
                }
            }
        }
        Node* RB_delete(int key){
            Node* z=search(key); //z is the node that we want to delete
            if(z==nullNode){
                cout<<"There is no node with this key";
                return;
            }
            Node* y;//y is the node that gets deleted from tree
            if(z->right==nullNode || z->left==nullNode){
                y=z;
            }else{
                
            }

            

        }
        void RB_delete_fixup(Node* x){

        }
        Node* search(int key){
            Node* currentNode=root;
            if(root==nullNode)return root;
            while(currentNode!=nullNode){
                if(currentNode->key==key)return currentNode;
                if(currentNode->key<key)currentNode=currentNode->right;
                else currentNode=currentNode->left;
            }
            return nullNode;
        }
        Node* create_node(int key){
            Node* newNode=new Node(key,red);
            newNode->right=newNode->left=newNode->parent=nullNode;
            return newNode;
        }
};



